contracts/parent-chain/espresso-migration/EspressoOspMigrationAction.sol

```diff
@@ -0,0 +1,102 @@
+
+    error AddressIsNotContract(address incorrectAddr);
+
+    error ChallengeManagerUpdated(address newChallengeManagerAddr);
+
+    error OspNotUpgraded(address oldOspAddress);
+
+    error WasmModuleRootNotUpdated(bytes32 oldWasmModuleRoot);
+
+    address public immutable newOspEntry;
+    bytes32 public immutable newWasmModuleRoot;
+    bytes32 public immutable currentWasmModuleRoot;
+    address public immutable currentOspEntry;
+    address public immutable rollup;
+    address public immutable proxyAdmin;
+
+
+    constructor(
+        address _newOspEntry,
+        bytes32 _newWasmModuleRoot,
+        address _currentOspEntry,
+        bytes32 _currentWasmModuleRoot,
+        address _rollup,
+        address _proxyAdmin
+    ){
+        if(_newWasmModuleRoot == bytes32(0)){
+            revert IncorrectWasmModuleRoot(_newWasmModuleRoot);
+        }
+
+        if(_currentWasmModuleRoot == bytes32(0)){
+            revert IncorrectWasmModuleRoot(_currentWasmModuleRoot);
+        }
+
+        if(!Address.isContract(_newOspEntry)){
+            revert AddressIsNotContract(_newOspEntry);
+        }
+
+        if(!Address.isContract(_currentOspEntry)){
+            revert AddressIsNotContract(_currentOspEntry);
+        }
+
+        if(!Address.isContract(_proxyAdmin)){
+            revert AddressIsNotContract(_proxyAdmin);
+        }
+
+        if(!Address.isContract(_rollup)){
+            revert AddressIsNotContract(_rollup);
+        }
+        newOspEntry = _newOspEntry;
+        newWasmModuleRoot = _newWasmModuleRoot;
+        currentOspEntry = _currentOspEntry;
+        currentWasmModuleRoot = _currentWasmModuleRoot;
+        rollup = _rollup;
+        proxyAdmin = _proxyAdmin;
+    }
+
+    function perform() external{
+        //Handle assertions in the perform function as we shouldn't be storing local state for delegated calls.
+        
+        // set the new challenge manager impl
+        TransparentUpgradeableProxy challengeManager =
+            TransparentUpgradeableProxy(payable(address(IRollupCore(rollup).challengeManager())));
+        address chalManImpl = ProxyAdmin(proxyAdmin).getProxyImplementation(challengeManager);
+        ProxyAdmin(proxyAdmin).upgradeAndCall(
+            challengeManager,
+            chalManImpl, // Use the rollups current challenge manager as we only need to upgrade the OSP
+            abi.encodeWithSelector(IChallengeManager.postUpgradeInit.selector, IOneStepProofEntry(newOspEntry), currentWasmModuleRoot, IOneStepProofEntry(currentOspEntry))
+        );
+        address postUpgradeChalManAddr = ProxyAdmin(proxyAdmin).getProxyImplementation(challengeManager);
+
+        if(postUpgradeChalManAddr != chalManImpl){
+            revert ChallengeManagerUpdated(postUpgradeChalManAddr);
+        }
+        IOneStepProofEntry newOsp = IChallengeManager(address(challengeManager)).osp();
+
+        if(newOsp != IOneStepProofEntry(newOspEntry) ){
```

Comment:
Is there a reason why we are storing `newOspEntry` in the constructor? If we store it in the constructor wont we have to deploy a new contract everytime we have to perform an upgrade?

Reply:
I do see your point that we would have to deploy a new version of the contract for each migration. I think that would be more in line with the approach intended for Arbitrum upgrades.

Typically these contracts are supposed to be instantiated with specific magic values for the constructor arguments that make it easier for the Arbitrum DAO to see exactly what would happen when the upgrade is ran. 

I might be wrong but wouldn't each rollup want a distinct OSP entry anyways?